### PR TITLE
Clippy in Regressions.

### PR DESCRIPTION
**Pull request by YoshikiTakashima**
### Description of changes: 

No change to kani behavior. Added `clippy` linter to regression suite. 

### Resolved issues:

Lack of clippy lint checking in kani regression script. #1373 will stay open to track actual clippy errors.

### Call-outs:
The current setup addresses the following clippy regressions.
- New clippy error/warning in a library that did not have them (thus not ignored)
- New class of clippy warnings/errors by code change

### Testing:

* How is this change tested? Run the regression script

* Is this a refactor change?

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: .github/workflows/format-check.yml
Comment (giltho) [inline .github/workflows/format-check.yml:39]: Can this specific configuration be put in a clippy.toml or something similar?
If I run cargo clippy on my machine, I want it to run just like on the CI, since this is a repo-wide decision, not just a CI decision.

And then, they can also be removed from the lib files

File: scripts/setup/macos-10.15/install_deps.sh
Comment (giltho) [inline scripts/setup/macos-10.15/install_deps.sh:13]: Why is jq required? I think you used it for the script that gathers data, but you didn't commit that script

**Review by giltho — COMMENTED** [review]
I'm not sure how clippy gets configured for a whole workspace, so I'm not sure if the first comment has a solution, the second one should probably be "fixed" before merging

File: .github/workflows/format-check.yml
Comment (YoshikiTakashima) [reply .github/workflows/format-check.yml:39]: @giltho I think that would be good as well. I'm less sure of the exact method to implement it. If you know of a working example, please send me the link.

A quick search suggests the method for doing that seems to be through `.cargo/config.toml`. https://github.com/emilk/egui/blob/e76c919c7e70c208c9a6209b9fe3369e7b6db99d/.cargo/config.toml

Let me prototype a little bit b/c I am not sure if this will have negative effects. I read **unconfirmed** rumors that this negatively effects incremental recompile.

File: scripts/setup/macos-10.15/install_deps.sh
Comment (YoshikiTakashima) [reply scripts/setup/macos-10.15/install_deps.sh:13]: Oops. Thanks for catching it. I will add the statistics gathering command

Comment (tedinski) [reply scripts/setup/macos-10.15/install_deps.sh:13]: If we're clean now, I'm not sure we need it. But at any rate we don't need to add jq to CI if it's not being used

Comment (YoshikiTakashima) [reply scripts/setup/macos-10.15/install_deps.sh:13]: We are not clean. This PR papers over the actual warning by turning off the lints. 

I believe this data should be tracked, but I agree you don't need to install jq on everything. Clippy check is only running for Ubuntu-20.04.

Comment (YoshikiTakashima) [reply scripts/setup/macos-10.15/install_deps.sh]: Addressed with the following:
stats print script: 37cd8030b4492a07630401e74e0ffdf70eb21abd
remove jq install from OSX: 7803971d1f646310f146c3658f284a552feb9af5

As of now, no stats will print b/c lints are turned off: 

P.S. Wait what, did GH just suggest a fix??? 
```suggestion
brew install universal-ctags wget jq
```

File: .github/workflows/format-check.yml
Comment (YoshikiTakashima) [reply .github/workflows/format-check.yml:39]: Update:
1. Incremental recompile issue can be mitigated.
2. However, .cargo/config.toml does not play nicely with the custom build command for crates under `library`.
```
error: failed to run custom build command for `kani-compiler v0.6.0 (/Users/ytakashl/Desktop/kani/kani-compiler)`

Caused by:
  process didn't exit successfully: `/Users/ytakashl/Desktop/kani/target/debug/build/kani-compiler-bfa3790a81ae900c/build-script-build` (exit status: 1)
  --- stdout
  cargo:rustc-link-arg-bin=kani-compiler=-Wl,-rpath,/Users/ytakashl/.rustup/toolchains/nightly-2022-07-19-aarch64-apple-darwin/lib
  cargo:rustc-link-arg-bin=kani-compiler=-Wl,-rpath,@loader_path/../toolchain/lib
  cargo:rerun-if-changed=../library/kani

  --- stderr
     Compiling kani_macros v0.6.0 (/Users/ytakashl/Desktop/kani/library/kani_macros)
  error: length comparison to zero
    --> library/kani_macros/src/lib.rs:42:13
     |
  42 |     assert!(attr.to_string().len() == 0, "#[kani::proof] does not take any arguments");
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is clearer and more explicit: `attr.to_string().is_empty()`
     |
     = note: `-D clippy::len-zero` implied by `-D warnings`
```

While there are only 4 warnings under `library/`, one of the warnings is a bit thorny b/c it asks us to define a safety doc comment for Invariant.

Comment (YoshikiTakashima) [reply .github/workflows/format-check.yml]: Resolved by adding `.cargo/config.toml`.

Warnings under `library/` are fixed since they appear not to be ignored.

**Review by celinval — COMMENTED** [review]
Thanks for doing this. I have 2 questions:
1. Is there a reason why you preferred `.cargo/config.toml` instead of `clippy.toml`.
2. How long does `cargo clippy` takes? I was wondering if we should enable this in our regression.

**Comment by YoshikiTakashima** [comment]
@celinval 

> 1. Is there a reason why you preferred `.cargo/config.toml` instead of `clippy.toml`.

I don't think that is possible, and keeping it so seems to be the stance from clippy devs: https://github.com/rust-lang/rust-clippy/issues/1313#issuecomment-633991593. The idea behind it seems to be that you would control all lints at once, not just clippy lints.

> 2. How long does `cargo clippy` takes? I was wondering if we should enable this in our regression.

It takes no longer than 2 compiles. This is the first job to finish.

**Comment by giltho** [comment]
LGTM ✅

**Review by giltho — APPROVED** [review]

**Comment by YoshikiTakashima** [comment]
Not sure how big the "code owner" set is but it still saids it waiting for review. @tedinski / @celinval

File: .cargo/config.toml
Comment (tedinski) [inline .cargo/config.toml:31]: appears above

File: .gitignore
Comment (tedinski) [inline .gitignore:28]: Huh?

File: scripts/setup/ubuntu/install_deps.sh
Comment (tedinski) [inline scripts/setup/ubuntu/install_deps.sh:29]: ?

Comment (celinval) [reply scripts/setup/ubuntu/install_deps.sh:29]: > @celinval
> 
> > 1. Is there a reason why you preferred `.cargo/config.toml` instead of `clippy.toml`.
> 
> I don't think that is possible, and keeping it so seems to be the stance from clippy devs: [rust-lang/rust-clippy#1313 (comment)](https://github.com/rust-lang/rust-clippy/issues/1313#issuecomment-633991593). The idea behind it seems to be that you would control all lints at once, not just clippy lints.
> 
> > 2. How long does `cargo clippy` takes? I was wondering if we should enable this in our regression.
> 
> It takes no longer than 2 compiles with `cargo build --workspace`. This is the first job to finish.

This is so confusing... That's totally what they recommend in their documentation: https://github.com/rust-lang/rust-clippy#configuration

File: .gitignore
Comment (YoshikiTakashima) [reply .gitignore:28]: 27 includes `*/.cargo`, 28 negates only top-level config.

I did that to minimize semantic diff, but I can just get rid of 27 as well.

File: scripts/setup/ubuntu/install_deps.sh
Comment (YoshikiTakashima) [reply scripts/setup/ubuntu/install_deps.sh:29]: @tedinski `jq` is required for generating the warning stats given at the bottom of .github/workflows/format-check.yml

Comment (YoshikiTakashima) [reply scripts/setup/ubuntu/install_deps.sh:29]: @celinval Yea, both are under the same section, which is not great.

I could not find a single example of allowing lints with `clippy.toml` so I went looking around and landed on that issue thread.

File: .cargo/config.toml
Comment (YoshikiTakashima) [reply .cargo/config.toml]: Fixed with 88a98da5fd200043b8bc0090b61c8d028db99ddf

File: .gitignore
Comment (YoshikiTakashima) [reply .gitignore:28]: Closing as this appears to be a question rather than a request.

File: scripts/setup/ubuntu/install_deps.sh
Comment (YoshikiTakashima) [reply scripts/setup/ubuntu/install_deps.sh:29]: Closing as this appears to be a question rather than a request.

File: .gitignore
Comment (tedinski) [reply .gitignore:28]: Does .cargo appear anywhere else? I don't think so, just removing the line seems best.

File: scripts/setup/ubuntu/install_deps.sh
Comment (tedinski) [reply scripts/setup/ubuntu/install_deps.sh:29]: I would put instally jq into that workflow, not as a general dependency.

**Review by tedinski — APPROVED** [review]

File: .gitignore
Comment (YoshikiTakashima) [reply .gitignore]: No effect on regressions. I will just get rid of line 27.

File: scripts/setup/ubuntu/install_deps.sh
Comment (YoshikiTakashima) [reply scripts/setup/ubuntu/install_deps.sh]: Moved in commit 455c42f492c79745e94ceaed92acf8cc0d27ed14

**Pull request merged**
